text/3716-target-modifiers.md

```diff
@@ -0,0 +1,514 @@
+require exact matches, then all CUs must agree because all CUs depend on
+`core`. The missed detection requires that we don't consider AB or BC
+incompatible. This could be an argument in favor of not allowing any mismatches
+with the shadow call stack sanitizer (which you never want to mix in practice
+anyway).
+
+The dynamic linking case is considered acceptable. Detecting it is out of scope
+of this RFC.
+
+## Name mangling
+
+It has been proposed that the modified target could be encoded in the name
+mangling scheme to help catch the two cases from [the previous
+section][not-caught]. However, this raises a bunch of open questions:
+
+1. It probably does not help catch the first case. Dynamic function calls
+   between the leaf modules wouldn't get caught, so that would require that one
+   of the leaf modules references a symbol defined by the other leaf module.
+   However, I find it hard to imagine this happening in the real world unless
+   the symbol is marked `#[no_mangle]`.
+2. Similarly, if two dynamic objects are compiled completely separately, they
+   probably do not reference each other through anything other than symbols
+   marked `#[no_mangle]`. While it could potentially identify a mismatch where
+   component A depends on component B, and component B is recompiled with a
+   different ABI while using the old version of A, this scenario is not
+   well-supported to start with because of Rust's unstable ABI.
+3. Some ABI-affecting flags only change the C ABI, but those symbols are
+   usually using `#[no_mangle]`.
+4. Do we really want to make our symbol names even longer?
+
+For the above reasons, name mangling is not proposed as a mechanism for
+detection for now. However, it could be a potential future addition.
+
+## Policy around flags that might not be ABI affecting
+
+Some flags have an unclear status where it is unclear whether it affects the
+ABI. For example, `-Zpatchable-function-entry` (which adds nop instructions
+before/after the function entrypoint) generally isn't considered to affect the
+ABI, but if combined with `-Zsanitizer=kcfi` then it does affect it since kcfi
+works by placing a hash of the function signature before the function
+entrypoint. Since `-Zsanitizer=kcfi` already needs to be a target modifier
+_anyway_, you could argue that `-Zpatchable-function-entry` doesn't need to be
+one.
+
+However, for these cases where we are uncertain, we take a conservative
+approach and mark them as a target modifier. It is not breaking to relax the
+rules in future releases.
+
+As for flags such as `-Cllvm-args` that can do basically anything, it may make
+more sense to just rename it to `-Cunsafe-llvm-args` rather than use the target
+modifier functionality.
+
+## Problems with mixing non-target-modifiers
+
+It's rather common to want a flag to be applied everywhere, even if it is not
+ABI breaking. This section discusses some of these cases.
```

Comment:
It's just some examples where the same compat check would be useful, even if mixing them is not unsound.

Reply:
But what does that mean for the RFC? Why is it part of the RFC?

Reply:
I mean, I think it's important to understand related cases. When I talked with C folks at the Linux Plumbers conference, they thought that the mitigation flags were *more* important to detect mismatches for than ABI issues. Because the mitigation flags fail silently, but ABI issues are loud.

Reply:
If that's the reason why we have this section, then the text should say so. :) Is the idea that in the future, maybe we want to make those flags part of the same check? Or have another check that works basically the same way but has a less scary name for the override flag?

Right now it's completely unclear why you are telling us the things you are telling us here.

Reply:
I have a little bit of that [here](https://github.com/Darksonn/rfcs/blob/target-modifiers/text/3716-target-modifiers.md#detection-for-non-target-modifiers), but I will clarify more.

Reply:
Yeah that's quite far removed from this section, it would help the reader to clarify this right at the point where I left this comment.

Reply:
Updated in https://github.com/rust-lang/rfcs/pull/3716/commits/50bee24c348689794c093e8f754dee9fb658d92f.